### ISSUE TITLE
Create VMs and assign a static IP

### ISSUE THREAD

**Issue by rthomson** [vmware/ansible-module-vcloud-director#15]
Thanks for supporting ansible users with these provisioning modules, VMware!

I'm wondering how one would go about creating multiple VMs within a vApp, each assigned their own static IP address using this project?

The only module that appears to support setting a static/manual IP address is `vcd_vapp`. However, I can't seem to use this module multiple times targeting the same vApp with different/separate VMs as the module tells me the "vApp is already present" when I run it after the first subsequent run where it creates the vApp and VM.

The `vcd_vapp_vm` module has `ip_allocation_mode` but no `ip_address` attribute.

Am I missing something or does this functionality not yet exist?

**Comment by carmineesposito73** [comment]
any news on this enhancement? it would be very useful
thanks

**Comment by sgutermann** [comment]
I second that. Would definately be usefull.

**Comment by mukultaneja** [comment]
@rthomson, please use vappvmnic module to assign static IPs to the VM and acknowledge if it works for you.

**Comment by carmineesposito73** [comment]
@rthomson @mukultaneja with the new `vcd_vapp_vm_nic` to create a vm with nic0 with a static address you have to:

- create the vm (in my use case I have a vm with nic0 with dhcp)
- add a vm nic (should be nic1)
- delete nic0
- add a vm nic (should be nic0)
- delete nic1

you cannot add a nic to a vm without any nics (the NetworkConnectionSection is not present in vms without nics)

**Comment by sgutermann** [comment]
@carmineesposito73 thank you for posting that. I though I was going crazy. I had to make a template with a nic connection in order to get creating VMs. I was not able to get a template running and adding a nic to it afterwards.

Funny enough that does work using the HTML5 Client. So IMO there must be a way.

**Comment by mukultaneja** [comment]
@all, I have updated vcd_vapp_vm_nic module with update functionality of VM NICs. As of now, you can switch between Manual to DHCP mode and update NIC's Ip Address in Manual mode. Update me if I have covered me this enhancement.

**Comment by carmineesposito73** [comment]
@mukultaneja, I tested the new module: i can switch between all allocation modes: DHCP, POOL and MANUAL and I can change IP address in MANUAL mode.
Changing the network is not covered, right?

**Comment by carmineesposito73** [comment]
here https://github.com/carmineesposito73/ansible-module-vcloud-director/blob/master/modules/vcd_vapp_vm_nic.py you can find my proposal on this modules:
* added nic_id parameter to create a nic with id chosen by the user (can be useful, for example, if you like to have, on different vms, the same nic id on a specific network, e.g nic_id = 9 on network "mgmt")
* update operation can be done even if network change is needed

**Comment by mukultaneja** [comment]
@carmineesposito73 network is also covered!

**Comment by carmineesposito73** [comment]
@mukultaneja, I had errors, I'll do more tests

**Comment by carmineesposito73** [comment]
@mukultaneja, I got an exception trying to change network.
```
Value of unknown type: <class 'pyvcloud.vcd.exceptions.BadRequestException'>, Status code: 400/BAD_REQUEST, Bad request  - Unexpected JAXB Exception  - cvc-complex-type.2.4.d: Invalid content was found starting with element 'network'. No child element is expected at this point. (request id: 573f3815-7141-4be9-b97e-0eb036b1805b)\r\n", 
```

To simplify I tried to change network of a DHCP nic, I've done this tests

```
  tasks:
  - name: update vm nic
    vcd_vapp_vm_nic:
        vm_name: "{{ my_vm }}"
        vapp: "{{ my_vapp }}"
        vdc: "{{ my_vdc }}"
        network: "{{ my_network }}"
        #ip_allocation_mode: "{{ my_ip_mode }}"
        nic_id: "{{ my_nic_id }}"
        operation: "update"
```
```
  tasks:
  - name: update vm nic
    vcd_vapp_vm_nic:
        vm_name: "{{ my_vm }}"
        vapp: "{{ my_vapp }}"
        vdc: "{{ my_vdc }}"
        network: "{{ my_network }}"
        ip_allocation_mode: "{{ my_ip_mode }}"
        nic_id: "{{ my_nic_id }}"
        operation: "update"
```

**Comment by carmineesposito73** [comment]
@mukultaneja, to make things working I've done these changes (probably you can do a better job)
```
diff --git a/modules/vcd_vapp_vm_nic.py b/modules/vcd_vapp_vm_nic.py
index f5ee765..a69db8d 100644
--- a/modules/vcd_vapp_vm_nic.py
+++ b/modules/vcd_vapp_vm_nic.py
@@ -259,19 +259,38 @@ class VappVMNIC(VcdAnsibleModule):
             raise EntityNotFoundException('Can\'t find the specified VM nic')
         nic_to_update = nic_indexs.index(nic_id)
 
-        if network:
-            nics.NetworkConnection[nic_to_update].network = network
-            response['changed'] = True
-
         if ip_allocation_mode:
             allocation_mode_element = E.IpAddressAllocationMode(ip_allocation_mode)
             nics.NetworkConnection[nic_to_update].IpAddressAllocationMode = allocation_mode_element
             response['changed'] = True
+        else:
+            # get the current ip allocation mode
+            ip_allocation_mode = str(nics.NetworkConnection[nic_to_update].IpAddressAllocationMode)
 
-        if ip_address:
+        if ip_allocation_mode == 'MANUAL' and ip_address and not network:
             nics.NetworkConnection[nic_to_update].IpAddress = E.IpAddress(ip_address)
             response['changed'] = True
 
+        if network:
+            if ip_allocation_mode in ('DHCP', 'POOL'):
+                network_element = E.NetworkConnection(
+                                E.NetworkConnectionIndex(nic_id),
+                                E.IsConnected(True),
+                                E.IpAddressAllocationMode(ip_allocation_mode),
+                                network=network)
+            else:
+                if not ip_address:
+                    raise Exception('IpAddress is missing.')
+                network_element = E.NetworkConnection(
+                                E.NetworkConnectionIndex(nic_id),
+                                E.IpAddress(ip_address),
+                                E.IsConnected(True),
+                                E.IpAddressAllocationMode(ip_allocation_mode),
+                                network=network)
+            #nics.NetworkConnection[nic_to_update].network = network
+            nics.NetworkConnection[nic_to_update] = network_element
+            response['changed'] = True
+
         if response['changed']:
             update_nic_task = self.client.put_resource(uri, nics, EntityType.NETWORK_CONNECTION_SECTION.value)
             self.execute_task(update_nic_task)
```

**Comment by mukultaneja** [comment]
@carmineesposito73, I have added a patch to update vm nic network. Please check once https://github.com/vmware/ansible-module-vcloud-director/commit/26ede9e5255492ba71cb848abd69db634e81e1f6

**Comment by carmineesposito73** [comment]
@mukultaneja, it works but I've found one more issue:
I cannot change allocation mode from DHCP to MANUAL (in this case two steps are required: DHCP -> POOL -> MANUAL).
This is the exception I got:
```
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "module_stderr": "Shared connection to localhost closed.\r\n", 
    "module_stdout": "/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\n/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\n/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\n/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\n/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\n/usr/local/lib/python3.6/dist-packages/urllib3/connectionpool.py:847: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings\r\n  InsecureRequestWarning)\r\nTraceback (most recent call last):\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/__main__.py\", line 340, in main\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/__main__.py\", line 163, in manage_operations\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/__main__.py\", line 276, in update_nic\r\n  File \"/usr/local/lib/python3.6/dist-packages/pyvcloud/vcd/client.py\", line 1041, in put_resource\r\n    objectify_results=objectify_results)\r\n  File \"/usr/local/lib/python3.6/dist-packages/pyvcloud/vcd/client.py\", line 861, in _do_request\r\n    response), _objectify_response(response, objectify_results))\r\n  File \"/usr/local/lib/python3.6/dist-packages/pyvcloud/vcd/client.py\", line 866, in _response_code_to_exception\r\n    raise BadRequestException(sc, request_id, objectify_response)\r\npyvcloud.vcd.exceptions.BadRequestException: Status code: 400/BAD_REQUEST, Bad request  - Unexpected JAXB Exception  - cvc-complex-type.2.4.d: Invalid content was found starting with element 'IpAddress'. No child element is expected at this point. (request id: c66719c0-57e5-4f11-a264-e658b82531ae)\r\n\r\nDuring handling of the above exception, another exception occurred:\r\n\r\nTraceback (most recent call last):\r\n  File \"/home/carmine/.ansible/tmp/ansible-tmp-1546945229.69-239004321790078/AnsiballZ_vcd_vapp_vm_nic.py\", line 113, in <module>\r\n    _ansiballz_main()\r\n  File \"/home/carmine/.ansible/tmp/ansible-tmp-1546945229.69-239004321790078/AnsiballZ_vcd_vapp_vm_nic.py\", line 105, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/home/carmine/.ansible/tmp/ansible-tmp-1546945229.69-239004321790078/AnsiballZ_vcd_vapp_vm_nic.py\", line 48, in invoke_module\r\n    imp.load_module('__main__', mod, module, MOD_DESC)\r\n  File \"/usr/lib/python3.6/imp.py\", line 235, in load_module\r\n    return load_source(name, filename, file)\r\n  File \"/usr/lib/python3.6/imp.py\", line 170, in load_source\r\n    module = _exec(spec, sys.modules[name])\r\n  File \"<frozen importlib._bootstrap>\", line 618, in _exec\r\n  File \"<frozen importlib._bootstrap_external>\", line 678, in exec_module\r\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/__main__.py\", line 352, in <module>\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/__main__.py\", line 346, in main\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/ansible_vcd_vapp_vm_nic_payload.zip/ansible/module_utils/basic.py\", line 2369, in fail_json\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/ansible_vcd_vapp_vm_nic_payload.zip/ansible/module_utils/basic.py\", line 2341, in _return_formatted\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/ansible_vcd_vapp_vm_nic_payload.zip/ansible/module_utils/basic.py\", line 522, in remove_values\r\n  File \"/tmp/ansible_vcd_vapp_vm_nic_payload_503i2a1n/ansible_vcd_vapp_vm_nic_payload.zip/ansible/module_utils/basic.py\", line 505, in _remove_values_conditions\r\nTypeError: Value of unknown type: <class 'pyvcloud.vcd.exceptions.BadRequestException'>, Status code: 400/BAD_REQUEST, Bad request  - Unexpected JAXB Exception  - cvc-complex-type.2.4.d: Invalid content was found starting with element 'IpAddress'. No child element is expected at this point. (request id: c66719c0-57e5-4f11-a264-e658b82531ae)\r\n", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}
```

**Comment by carmineesposito73** [comment]
The issue DHCP to MANUAL is present with VM powered off, when the IpAddress field is missing and cannot be modified, this is the output of read command with VM powered off
```
    "msg": {
        "0": {
            "IpAddressAllocationMode": "DHCP", 
            "IsConnected": "True", 
            "MACAddress": "00:50:56:01:3c:b0", 
            "NetworkAdapterType": "VMXNET3", 
            "NetworkConnectionIndex": "0"
        }
    }
```

**Comment by carmineesposito73** [comment]
I will open a specific issue on that

**Comment by rthomson** [comment]
Thanks for all the work on this folks! I've been sidetracked from my initial work with ansible and vCD so I haven't had the chance to test the new module yet but I will try to do soon.

**Comment by carmineesposito73** [comment]
@mukultaneja here https://github.com/carmineesposito73/ansible-module-vcloud-director/commit/611a45937c8923644cd6e67474f15b37cf6d0437 you can find my last commit with some differences/improvements on `vcd_vapp_vm_nic.py`

Differences beetwen my module and the official one:
* function `add_nic`: nic_id can be selected, if nic_id is already present nothing happens (warning message), if nic_id is not specified, automatic nic number is used (same as official function)
@rthomson do you think it's useful to have the option to chose the nic_id of a new nic?
* function `update_nic`: 1. resolved #34 - 2. warning message is used (exception not raised) when update operation is requested on a nic_id not present
* function `read_nics`: "network" information added
